python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,151 @@
+        ret = super()._from_host_memory(data)
+        ret._finalize_init()
+        return ret
+
+    def _is_shared(self):
+        """
+        Return `True` if `self`'s memory is shared with other columns.
+        """
+        return len(self.__class__._instances[(self.ptr, self.size)]) > 1
+
+    def copy(self, deep: bool = True):
+        """
+        Return a copy of Buffer.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            If True, returns a deep-copy of the underlying Buffer data.
+            If False, returns a shallow-copy of the Buffer pointing to
+            the same underlying data.
+
+        Returns
+        -------
+        Buffer
+        """
+        if not deep:
+            copied_buf = CopyOnWriteBuffer.__new__(CopyOnWriteBuffer)
+            copied_buf._ptr = self._ptr
+            copied_buf._size = self._size
+            copied_buf._owner = self._owner
+            copied_buf._finalize_init()
+            return copied_buf
+        else:
+            owner_copy: rmm.DeviceBuffer = copy.copy(self._owner)
+            return self._from_device_memory(
+                cuda_array_interface_wrapper(
+                    ptr=owner_copy.ptr,
+                    size=owner_copy.size,
+                    owner=owner_copy,
+                )
+            )
+
+    @property
+    def _cuda_array_interface_readonly(self) -> dict:
+        """
+        Internal Implementation for the CUDA Array Interface without
+        triggering a deepcopy.
+        """
+        return {
+            "data": (self.ptr, True),
+            "shape": (self.size,),
+            "strides": None,
+            "typestr": "|u1",
+            "version": 0,
+        }
+
+    @property
+    def __cuda_array_interface__(self) -> dict:
+        # Detach if there are any weak-references.
+
+        # Mark the Buffer as ``zero_copied=True``,
+        # which will prevent any copy-on-write
+        # mechanism post this operation.
+        # This is done because we don't have any
+        # control over knowing if a third-party library
+        # has modified the data this Buffer is
+        # pointing to.
+        self._unlink_shared_buffers(zero_copied=True)
+
+        result = self._cuda_array_interface_readonly
+        result["data"] = (self.ptr, False)
+        return result
+
+    def _unlink_shared_buffers(self, zero_copied=False):
+        """
+        Unlinks a Buffer if it is shared with other buffers(i.e., weak references exist) by making
+        a true deep-copy.
+        """
+        if not self._zero_copied and self._shallow_copied():
+            # make a deep copy of existing DeviceBuffer
+            # and replace pointer to it.
+            current_buf = rmm.DeviceBuffer(ptr=self.ptr, size=self.size)
+            new_buf = current_buf.copy()
+            self._ptr = new_buf.ptr
+            self._size = new_buf.size
+            self._owner = new_buf
+            self._finalize_init()
```

Comment:
Why are we not just implementing `CopyOnWriterBuffer.__init__`? Do you need finer-grained control over what is going on?

Reply:
Thanks.